core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -254,18 +255,36 @@ class ReplicaFetcherThread(name: String,
       // We will not include a replica in the fetch request if it should be throttled.
       if (fetchState.isReadyForFetch && !shouldFollowerThrottle(quota, fetchState, topicPartition)) {
         try {
-          val logStartOffset = this.logStartOffset(topicPartition)
-          builder.add(topicPartition, new FetchRequest.PartitionData(
-            fetchState.fetchOffset, logStartOffset, fetchSize, Optional.of(fetchState.currentLeaderEpoch)))
+          // A fetch needs to be used to update the session only if:
```

Comment:
It's too bad this loop is still O(n) as a result of the `shouldFollowerThrottle`, but at least we've minimized the amount of work to be done inside the loop.